zebra/zebra_wrap_script.c

```diff
@@ -0,0 +1,1107 @@
+ *     0     0     DROP      all --  any   any anywhere  anywhere
+ *          match-set match0x55f44       (## line 2)
+ * json_obj_list : the json structure mapped to the output, ranked with line nb
+ * - ex: above dump gives following
+ * { "2":{"pkts":"0","bytes":"0","target":"MARK","prot":"all", \
+ *           "opt":"--","in":"any",..}}
+ */
+static int zebra_wrap_script_rows(const char *script,
+			   int begin_at_line,
+			   struct json_object *json_obj_list)
+{
+	FILE *fp;
+	char data[DATA_LEN];
+	char *current_str = NULL;
+	int nb_entries = 0;
+	int line_nb = 0, i;
+	struct item_list item[ITEM_MAXIMUM];
+
+	/* initialise item list
+	 */
+	for (i = 0; i < ITEM_MAXIMUM; i++)
+		memset(&item[i], 0, sizeof(struct item_list));
+	if (IS_ZEBRA_DEBUG_KERNEL_MSGDUMP_SEND)
+		zlog_debug("SCRIPT : %s", script);
+	fp = popen(script, "r");
+	if (!fp) {
+		zlog_err("SCRIPT: error calling %s", script);
+		return -1;
+	}
+	do {
+		json_object *json_obj = NULL;
+		char separator[5];
+
+		memset(separator, 0, sizeof(separator));
+		separator[0] = ' ';
+		memset(data, 0, DATA_LEN);
+		current_str = fgets(data, DATA_LEN, fp);
+		if (current_str) {
+			/* data contains the line
+			 */
+			current_str = data;
+			if (zebra_wrap_debug & SCRIPT_DEBUG)
+				zlog_debug("SCRIPT : [%d/%d] %s",
+					   line_nb,
+					   (int)strlen(current_str),
+					   current_str);
+			if ((strlen(current_str) <= 1) ||
+			    line_nb < begin_at_line) {
+				line_nb++;
+				continue;
+			}
+			if (line_nb > begin_at_line)
+				json_obj = json_object_new_object();
+			else
+				json_obj = NULL;
+			if (handle_field_line_row(json_obj, current_str,
+						  item, separator) < 0)
+				return -1;
+			if (json_obj) {
+				char line[10];
+
+				snprintf(line, sizeof(line), "%d", nb_entries);
+				json_object_object_add(json_obj_list,
+						       line, json_obj);
+				nb_entries++;
+			}
+			line_nb++;
+		}
+	} while (current_str != NULL);
+	/* free item list
+	 */
+	for (i = 0; i < ITEM_MAXIMUM; i++) {
+		if (item[i].name) {
+			XFREE(MTYPE_TMP, item[i].name);
+			item[i].name = NULL;
+		}
+	}
+	if (pclose(fp))
+		zlog_err("SCRIPT: error closing stream with %s", script);
+	return 0;
+}
+
+static int zebra_wrap_script_call_only(const char *script)
+{
+	FILE *fp;
+
+	if (IS_ZEBRA_DEBUG_KERNEL_MSGDUMP_SEND)
+		zlog_debug("SCRIPT : %s", script);
+	fp = popen(script, "r");
+	if (!fp) {
+		zlog_err("SCRIPT: error calling %s", script);
+		return -1;
+	}
+	if (pclose(fp)) {
+		zlog_err("SCRIPT: error with %s: closing stream (errno %u)",
+			 script, errno);
+		return -1;
+	}
+	return 0;
+}
+
+/* convert string <NUM>[K,M,G] into int64 value
+ * last letter of the word is a multiplier
+ * remove it and apply atoll
+ */
+static void zebra_wrap_script_convert_stat(const char *from, uint64_t *to,
+					   uint64_t multiplier)
+{
+	char buff_tmp[64];
+	char *ptr;
+
+	if (!from) {
+		*to = 0;
+		return;
+	}
+	strncpy(buff_tmp, from, sizeof(buff_tmp));
```

Comment:
`strncpy` does not null terminate if there is not enough space.

Reply:
I kept one char so that I will be able to mark null char , just after.